File: 6-zigzag-conversion/6-zigzag-conversion.js
/**
 * @param {string} s
 * @param {number} numRows
 * @return {string}
 */

var convert = function (s, numRows) {
    if (numRows === 1) return s;

    // 1. create array of length numRows
    const rows = new Array(numRows).fill('');

    // 2. newBase -> (numRows - 1) * 2
    const newBase = (numRows - 1) * 2;

    // 3. create lookup table
    const newIdxMods = new Array(newBase);
    for (let i = 0; i < newBase; i++) {
        newIdxMods[i] = i <= newBase / 2 ? i : newBase - i;
    }
    
    // 4. iterate through s, (ch, i)
    for (let i = 0; i < s.length; i++) {
        const ch = s[i];
        const iMod = i % newBase;
        rows[newIdxMods[iMod]] += ch;
    }

    return rows.join('');
};

File: 6-zigzag-conversion/6-zigzag-conversion.py
class Solution:
    def convert(self, s: str, numRows: int) -> str:
        if numRows == 1:
            return s
        rows = [''] * numRows
        newBase = (numRows - 1) * 2
        
        newIdxMods = [0] * newBase
        for i in range(newBase):
            newIdxMods[i] = i if i <= newBase / 2 else newBase - i
        
        for i, ch in enumerate(s):
            iMod = i % newBase
            rows[newIdxMods[iMod]] += ch
            
        return ''.join(rows)